### DISCUSSION TITLE
FrankenPHP Worker Runtime vs Classic Symfony: Filters Not Applying

### DISCUSSION THREAD

**Discussion by Selanas** [php/frankenphp#1263] [Q&A]
Hi everyone,

I need some help understanding a pretty weird problem I’m encountering with FrankenPHP’s worker mode and the FrankenPHP Runtime.

**Premises:**
Both environments run the exact same codebase, built into the same container image. The environment configurations are the same (except for the runtime runner). The Caddy server is the same, and so is the Caddy config. Everything is identical except for the running context. No cache issues, no gateway troubles—I’ve triple-checked all of that.

The first environment runs locally with the classic Symfony index.php. The second runs in worker mode using the FrankenPHP Runtime.

**Problem:**
When I run the exact same request, only on the worker mode run, my filters seem to be ignored. I go from 1 query to 63 queries (just some bad eager loading; the exact number isn’t what’s important here). The filters are ignored, and I’m left with a simple getCollection result.

All parameters are correctly passed in the trace, and I’ve tried both dev and prod mode. As soon as I redeploy without the worker mode, everything works fine again.

![image](https://github.com/user-attachments/assets/9e1de990-444e-4729-a9af-cb916e7fc107)

I’ve been stuck on this for hours, and I could really use a more distant/expert perspective on the matter. Any ideas to help me understand this problem?

Extra informations:

- The project is built around API Platform.
- The architecture follows Domain-Driven Design.
- We’ve re-implement the API Platform’s out-of-the-box filters with a custom providers using `#[AutowireIterator(tag: 'api_platform.doctrine.orm.query_extension.collection')].`
- A switch selects the correct kernel using `$context['APP_ID']`. (should'nt have any inpact at this point i think)
- No mercure hub or Vulcain at this point of the project

**Comment by withinboredom** [comment]
This seems suspiciously like some part of the request is from another request -- perhaps a dependency being injected isn't being reset after a request?

**Reply by Selanas** [reply]
I’m experiencing the same issue, even with freshly built containers. The strangest part is that I have real difficulties reproducing the problem locally.

Profiler reviews are almost identical across both environments. The configuration is exactly the same, _api_query_parameters are the same... but the query results are incoherent. (and some azure gateway headers)

Here’s a screenshot for reference:
![image](https://github.com/user-attachments/assets/bfcb65c3-65fc-44e1-8cec-f4a9f8959827)

![image](https://github.com/user-attachments/assets/959c3ca9-47ef-4d4f-b55b-844862884f49)
![image](https://github.com/user-attachments/assets/62d9bb80-273e-4a2e-91bb-4149cd96af95)
![image](https://github.com/user-attachments/assets/d8e6af59-45cc-45fa-ab72-1c56ce06cd83)


I’ve tried some exploratory debugging with Xdebug locally, but I’m more limited when it comes to the running environment. If you or anyone else has suggestions on what kind of probes I could set to investigate further, your input would be greatly appreciated!

ps : Wishing everyone reading this a wonderful holiday season and happy celebrations! 🎄🎉


ps 2 : How api platform filter are reconciliated : 

```php
abstract class AbstractCollectionProvider implements ProviderInterface
{
    use LinksHandlerLocatorTrait;
    use LinksHandlerTrait;

    public function __construct(
        private readonly IriConverterInterface $iriConverter,
        ManagerRegistry $managerRegistry,
        ResourceMetadataCollectionFactoryInterface $resourceMetadataCollectionFactory,
        #[AutowireLocator('api_platform.doctrine.orm.links_handler', indexAttribute: 'key')]
        ?ContainerInterface $handleLinksLocator = null,
        #[AutowireIterator(tag: 'api_platform.doctrine.orm.query_extension.collection')]
        private readonly iterable $collectionExtensions = [],
    ) {
        $this->resourceMetadataCollectionFactory = $resourceMetadataCollectionFactory;
        $this->handleLinksLocator = $handleLinksLocator;
        $this->managerRegistry = $managerRegistry;
    }

[...]

    public function provide(Operation $operation, array $uriVariables = [], array $context = []): array|Paginator
    {
        $entityClass = $this->getEntityClass();
        $resourceClass = $operation->getClass();
        $manager = $this->managerRegistry->getManagerForClass($entityClass);
        $repository = $manager->getRepository($entityClass);

        $queryBuilder = $repository->createQueryBuilder('o');
        $queryNameGenerator = new QueryNameGenerator();

        if ($handleLinks = $this->getLinksHandler($operation)) {
            $handleLinks($queryBuilder, $uriVariables, $queryNameGenerator, ['entityClass' => $entityClass, 'operation' => $operation] + $context);
        } else {
            $this->handleLinks($queryBuilder, $uriVariables, $queryNameGenerator, $context, $entityClass, $operation);
        }

        foreach ($this->collectionExtensions as $extension) {
            $extension->applyToCollection($queryBuilder, $queryNameGenerator, $entityClass, $operation, $context);

            if ($extension instanceof QueryResultCollectionExtensionInterface && $extension->supportsResult($entityClass, $operation, $context)) {
                return $this->fromModel($resourceClass, $context, $extension->getResult($queryBuilder, $entityClass, $operation, $context));
            }
        }

        return $this->fromModel($resourceClass, $context, $queryBuilder->getQuery()->getResult());
    }
```

**Reply by dunglas** [reply]
Regarding the inconsistency, be sure to set the number of workers to 1 (by default it's 2 times the number of CPUs). Otherwise other (fresh) workers could handle the request.

**Reply by Selanas** [reply]
Thank you for your suggestion! I tried reducing the workers to 1 with the following configuration:

```plaintext
worker {
    file ./public/index.php
    num 1
    env APP_RUNTIME Runtime\FrankenPhpSymfony\Runtime
}
```

After testing and reviewing the worker logs, I’m still encountering the same issue. it seems like filters are being ignored when applying extension filters, resulting in the entire collection being returned instead of the filtered, unique result.
The only anomaly I observed in the logs is the -2 ready workers, but I don't have enough understanding to determine if this is related to the problem or just an expected metric.


For more intels, the server metrics and caddyfile config, meaby i just doing something wrong with that

```
caddy_reverse_proxy_upstreams_healthy{upstream="my-service-urll:8080"} 1

frankenphp_busy_threads 1
frankenphp_public_index_php_busy_workers 0
frankenphp_public_index_php_ready_workers -2
frankenphp_public_index_php_total_workers 1
frankenphp_public_index_php_worker_crashes 1
frankenphp_public_index_php_worker_request_count 45
frankenphp_public_index_php_worker_request_time 3311.8580419039995
frankenphp_public_index_php_worker_restarts 2
frankenphp_total_threads 8

go_build_info{checksum="",path="github.com/dunglas/frankenphp/caddy",version="(devel)"} 1
go_gc_duration_seconds{quantile="0"} 6.1317e-05
go_gc_duration_seconds{quantile="0.25"} 0.000105289
go_gc_duration_seconds{quantile="0.5"} 0.000133281
go_gc_duration_seconds{quantile="0.75"} 0.000228301
go_gc_duration_seconds{quantile="1"} 0.000610676
go_gc_duration_seconds_sum 0.002199875
go_gc_duration_seconds_count 12
go_gc_gogc_percent 100
go_gc_gomemlimit_bytes 9.223372036854776e+18
go_goroutines 27
go_info{version="go1.23.4"} 1
go_memstats_alloc_bytes 7.61416e+06
go_memstats_alloc_bytes_total 2.0372248e+07
go_memstats_buck_hash_sys_bytes 1.45911e+06
go_memstats_frees_total 145317
go_memstats_gc_sys_bytes 3.529472e+06
go_memstats_heap_alloc_bytes 7.61416e+06
go_memstats_heap_idle_bytes 5.210112e+06
go_memstats_heap_inuse_bytes 1.0518528e+07
go_memstats_heap_objects 24362
go_memstats_heap_released_bytes 4.5056e+06
go_memstats_heap_sys_bytes 1.572864e+07
go_memstats_last_gc_time_seconds 1.7351652803536391e+09
go_memstats_mallocs_total 169679
go_memstats_mcache_inuse_bytes 4800
go_memstats_mcache_sys_bytes 15600
go_memstats_mspan_inuse_bytes 140800
go_memstats_mspan_sys_bytes 163200
go_memstats_next_gc_bytes 1.5567696e+07
go_memstats_other_sys_bytes 1.211378e+06
go_memstats_stack_inuse_bytes 1.048576e+06
go_memstats_stack_sys_bytes 1.048576e+06
go_memstats_sys_bytes 2.3155976e+07
go_sched_gomaxprocs_threads 4
go_threads 19

mercure_subscribers_connected 0
mercure_subscribers_total 0
mercure_updates_total 0

process_cpu_seconds_total 5.42
process_max_fds 1.048576e+06
process_network_receive_bytes_total 3.50156506e+08
process_network_transmit_bytes_total 9.249008e+06
process_open_fds 16
process_resident_memory_bytes 1.76361472e+08
process_start_time_seconds 1.73516456536e+09
process_virtual_memory_bytes 3.106308096e+09
process_virtual_memory_max_bytes 1.8446744073709552e+19

promhttp_metric_handler_requests_in_flight 1
promhttp_metric_handler_requests_total{code="200"} 0
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0

```

The caddy file running the app

```
{
        {$CADDY_GLOBAL_OPTIONS}

		log {
			output stdout
			format json
		}

		servers {
			timeouts {
				read_body 10s
				read_header 10s
				write 15s
				idle 2m
			}
		}

        frankenphp {
                {$FRANKENPHP_CONFIG}
        }

        order php_server before file_server

        {$CADDY_EXTRA_CONFIG}

        auto_https off
		admin off
}

http://{$CONTAINER_APP_HOSTNAME}:8080, http://{$SERVER_NAME}:8080 {

        log {
                format filter {
                        wrap json
                        fields {
                                uri query {
                                        replace authorization REDACTED
                                }
                        }
                }
        }

        metrics /metrics

        root * /app/public

        encode {
                zstd
                br
                gzip

                match {
                        header Content-Type text/*
                        header Content-Type application/json*
                        header Content-Type application/javascript*
                        header Content-Type application/xhtml+xml*
                        header Content-Type application/atom+xml*
                        header Content-Type application/rss+xml*
                        header Content-Type image/svg+xml*
                        header Content-Type application/ld+json*
                }
        }

        {$CADDY_SERVER_EXTRA_DIRECTIVES}

        header ?Permissions-Policy "browsing-topics=()"

        php_server

		# Ajout des en-têtes X-Forwarded-* pour le proxy inversé.
		reverse_proxy / {
			to {$CONTAINER_APP_HOSTNAME}:8080
			header_up X-Real-IP {http.request.header.X-Forwarded-For}
			header_up Host {http.request.header.X-Forwarded-Host}
			header_up X-Forwarded-Proto {http.request.header.X-Forwarded-Proto}
		}

}



```
